### PR TITLE
[MNIST] Implement minibatch gradient descent and print accuracy.

### PR DESCRIPTION
**Pull request by dan-zheng**
* Implement minibatch gradient descent.
  * TODO: Randomly sample minibatches using TensorFlow dataset APIs.
* Print accuracy (rather than loss) to clearly show that the model is improving.

Example output:
```
Reading data.
Constructing data tensors.
Begin training for 20 epochs.
Accuracy: 53172/60000 (0.8862)
Accuracy: 55651/60000 (0.92751664)
Accuracy: 56104/60000 (0.93506664)
Accuracy: 56390/60000 (0.93983334)
Accuracy: 56576/60000 (0.9429333)
Accuracy: 56810/60000 (0.9468333)
Accuracy: 56968/60000 (0.94946665)
Accuracy: 57105/60000 (0.95175)
Accuracy: 57228/60000 (0.9538)
Accuracy: 57358/60000 (0.95596665)
Accuracy: 57347/60000 (0.9557833)
Accuracy: 57357/60000 (0.95595)
Accuracy: 57367/60000 (0.9561167)
Accuracy: 57540/60000 (0.959)
Accuracy: 57545/60000 (0.9590833)
Accuracy: 57610/60000 (0.9601667)
Accuracy: 57657/60000 (0.96095)
Accuracy: 57632/60000 (0.9605333)
Accuracy: 57694/60000 (0.9615667)
Accuracy: 57722/60000 (0.96203333)
Accuracy: 57816/60000 (0.9636)
```

### CODE REVIEW

File: MNIST/MNIST.swift [hunk 1]
```diff
@@ -33,14 +33,14 @@ func readMNIST(imagesFile: String, labelsFile: String) -> (images: Tensor<Float>
...
-    var w2 = Tensor<Float>(randomUniform: [30, 10])
+    var w1 = Tensor<Float>(randomUniform: [784, 30]) * 0.1
+    var w2 = Tensor<Float>(randomUniform: [30, 10]) * 0.1
     var b1 = Tensor<Float>(zeros: [1, 30])
     var b2 = Tensor<Float>(zeros: [1, 10])
 }
 
 /// Train a MNIST classifier for the specified number of epochs.
-func train(_ parameters: inout MNISTParameters, epochCount: Int) {
+func train(_ parameters: inout MNISTParameters, epochCount: Int32) {
```
Comment (dan-zheng) [inline MNIST/MNIST.swift]: Changing `epochCount` to type `Int32` is necessary to work around [SR-8419](https://bugs.swift.org/browse/SR-8419).

File: MNIST/MNIST.swift [hunk 2]
```diff
@@ -33,14 +33,14 @@ func readMNIST(imagesFile: String, labelsFile: String) -> (images: Tensor<Float>
 
 /// Parameters of an MNIST classifier.
 struct MNISTParameters : ParameterAggregate {
-    var w1 = Tensor<Float>(randomUniform: [784, 30])
-    var w2 = Tensor<Float>(randomUniform: [30, 10])
+    var w1 = Tensor<Float>(randomUniform: [784, 30]) * 0.1
```
Comment (rxwei) [inline MNIST/MNIST.swift]: Consider switching to random normal?

File: MNIST/MNIST.swift [hunk 3]
```diff
@@ -54,45 +54,68 @@ func train(_ parameters: inout MNISTParameters, epochCount: Int) {
...
     // Hyper-parameters.
+    let minibatchSize: Int32 = 10
     let learningRate: Float = 0.2
     var loss = Float.infinity
 
     // Training loop.
     print("Begin training for \(epochCount) epochs.")
 
+    func minibatch<Scalar>(_ x: Tensor<Scalar>, index: Int32) -> Tensor<Scalar> {
+      return x[index*minibatchSize..<(index+1)*minibatchSize]
```
Comment (rxwei) [inline MNIST/MNIST.swift]: How about:

let start = index*minibatchSize
return x[start..<start+minibatchSize]

Shorter, and one fewer multiplication.

File: MNIST/MNIST.swift (hunk 2 above)
Comment (dan-zheng) [reply MNIST/MNIST.swift]: Good idea! Done in https://github.com/tensorflow/swift-models/pull/21/commits/c51346b9a40cb518bb11c1dc556fea6fbd157e7f.
Random normalization initialization seems to perform a fair bit better than random uniform.

File: MNIST/MNIST.swift (hunk 3 above)
Comment (dan-zheng) [reply MNIST/MNIST.swift]: Done in https://github.com/tensorflow/swift-models/pull/21/commits/c51346b9a40cb518bb11c1dc556fea6fbd157e7f.

File: MNIST/MNIST.swift [hunk 4]
```diff
@@ -54,45 +54,69 @@ func train(_ parameters: inout MNISTParameters, epochCount: Int) {
...
+
+            // Calculate the sigmoid-based cross-entropy loss.
+            // TODO: Use softmax-based cross-entropy loss instead. Sigmoid
+            // cross-entropy loss treats class labels as independent, which is
+            // unnecessary for single-label classification tasks like MNIST.
+            // Sigmoid cross-entropy formula from:
+            // https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits
+            let part1 = max(predictions, 0) - predictions * labels
+            let part2 = log(1 + exp(-abs(predictions)))
+            loss = ((part1 + part2).sum(squeezingAxes: 0, 1) / batchSize).scalarized()
```
Comment (dan-zheng) [inline MNIST/MNIST.swift]: Note: `loss` is effectively unused, now that it's no longer being printed.

Comment (dan-zheng) [reply MNIST/MNIST.swift]: Done in https://github.com/tensorflow/swift-models/pull/21/commits/4275fdd3aaab19b0c027e43fbdcd7702e5dd238c.

**Comment by dan-zheng** [comment]
Indeed, the new MNIST does not work with the latest toolchain (7/24), failing with the error:
```
SESE FIXME: Imperfect loop exits not handled yet!
UNREACHABLE executed at /Library/Caches/swift-build/swift/lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:802!
```

I'll check this PR into a new branch. How does this new branch hierarchy sound:
* Rename the current `master` branch to `stable`. Code on the `stable` branch should always work with the last released toolchains. Change `stable` to be the "default" GitHub branch.
* Create a new branch `master`, which is for development and works with [swift trunk](https://github.com/apple/swift/tree/tensorflow). As new toolchains get released, push `master` to `stable`.

**Comment by rxwei** [comment]
SGTM.

**Comment by dan-zheng** [comment]
@rxwei Thanks for setting "stable" as the default branch.
"master" will now be used for development. Merging.

File: MNIST/MNIST.swift (hunk 2 above)
Comment (dan-zheng) [reply MNIST/MNIST.swift]: For the record: initializing `w1` and `w2` with zeros (just like the [TF Python beginner tutorial](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/tutorials/mnist/mnist_softmax.py)) works too. The model doesn't converge as quickly though.
